plugins/modules/vm_list_group_by_clusters.py

```diff
@@ -0,0 +1,191 @@
+                "hw_folder": "/DC0/vm/Discovered virtual machine",
+                "hw_guest_full_name": null,
+                "hw_guest_ha_state": null,
+                "hw_guest_id": null,
+                "hw_interfaces": [
+                    "eth0"
+                ],
+                "hw_is_template": false,
+                "hw_memtotal_mb": 1024,
+                "hw_name": "ubuntu_t",
+                "hw_power_status": "poweredOff",
+                "hw_processor_count": 1,
+                "hw_product_uuid": "4207072c-edd8-3bd5-64dc-903fd3a0db04",
+                "hw_version": "vmx-13",
+                "instance_uuid": "5007769d-add3-1e12-f1fe-225ae2a07caf",
+                "ipv4": null,
+                "ipv6": null,
+                "module_hw": true,
+                "snapshots": [],
+                "tags": [
+                    "backup"
+                ],
+                "vnc": {},
+                "moid": "vm-42",
+                "vimref": "vim.VirtualMachine:vm-42"
+            }]
+        }
+    }
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.vmware.vmware.plugins.module_utils.vmware_rest_client import VmwareRestClient
+
+
+class VmwareVMList(VmwareRestClient):
+    def __init__(self, module):
+        super(VmwareVMList, self).__init__(module)
+        self.module = module
+        self.params = module.params
+        self.detailed_vms = self.params['detailed_vms']
+
+    def get_all_clusters(self):
+        return self.api_client.vcenter.Cluster.list()
+
+    def get_all_folders(self):
+        folder_filter_spec = self.api_client.vcenter.Folder.FilterSpec(
+            type=self.api_client.vcenter.Folder.Type.VIRTUAL_MACHINE)
+        return self.api_client.vcenter.Folder.list(folder_filter_spec)
+
+    def get_all_vms_in_cluster_and_folder(self, cluster, folder):
+        vms_filter = self.api_client.vcenter.VM.FilterSpec(clusters=set([cluster]),
+                                                           folders=set([folder]))
+        return self.api_client.vcenter.VM.list(vms_filter)
+
+    def get_vm_detailed(self, vm):
+        return self.api_client.vcenter.VM.get(vm=vm)
+
+    def get_vm_list_group_by_clusters(self):
+        clusters = self.get_all_clusters()
+        folders = self.get_all_folders()
+
+        result_dict = {}
+        for cluster in clusters:
+            vm_list_group_by_folder_dict = {}
+            empty = True
+            for folder in folders:
+                vms = self.get_all_vms_in_cluster_and_folder(cluster.cluster, folder.folder)
+                vms_detailed = []
+                for vm in vms:
+                    if self.detailed_vms:
+                        vm = self.get_vm_detailed(vm=vm.vm)
+
+                    vms_detailed.append(self._vvars(vm))
+
+                if vms_detailed:
+                    empty = False
+                    vm_list_group_by_folder_dict[folder.name] = vms_detailed
+
+            if not empty:
```

Comment:
If I understand correctly you can get rid of the `empty` variable, and just check if `vm_list_group_by_folder_dict` is empty.